app/grandchallenge/algorithms/views.py

```diff
@@ -321,3 +356,119 @@ class ResultViewSet(ReadOnlyModelViewSet):
     serializer_class = ResultSerializer
     permission_classes = [DjangoObjectPermissions]
     filter_backends = [ObjectPermissionsFilter]
+
+
+class AlgorithmPermissionRequestCreate(
+    UserIsNotAnonMixin, SuccessMessageMixin, CreateView
+):
+    model = AlgorithmPermissionRequest
+    fields = ()
+
+    @property
+    def algorithm(self):
+        return Algorithm.objects.get(slug=self.kwargs["slug"])
+
+    def get_success_url(self):
+        return self.algorithm.get_absolute_url()
+
+    def get_success_message(self, cleaned_data):
+        return self.object.status_to_string()
+
+    def form_valid(self, form):
+        form.instance.user = self.request.user
+        form.instance.algorithm = self.algorithm
+        try:
+            redirect = super().form_valid(form)
+            return redirect
+
+        except ValidationError as e:
+            form._errors[NON_FIELD_ERRORS] = ErrorList(e.messages)
+            return super().form_invalid(form)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        permission_request = AlgorithmPermissionRequest.objects.filter(
+            algorithm=self.algorithm, user=self.request.user
+        ).first()
+        context.update(
+            {
+                "permission_request": permission_request,
+                "algorithm": self.algorithm,
+            }
+        )
+        return context
+
+
+class AlgorithmPermissionRequestList(ObjectPermissionRequiredMixin, ListView):
+    model = AlgorithmPermissionRequest
+    permission_required = (
+        f"{Algorithm._meta.app_label}.change_{Algorithm._meta.model_name}"
+    )
+    raise_exception = True
+
+    @property
+    def algorithm(self):
+        return Algorithm.objects.get(slug=self.kwargs["slug"])
+
+    def get_permission_object(self):
+        return self.algorithm
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        queryset = queryset.filter(algorithm=self.algorithm)
+        return queryset
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context.update({"algorithm": self.algorithm})
+        return context
+
+
+class AlgorithmPermissionRequestUpdate(SuccessMessageMixin, UpdateView):
```

Comment:
Is this missing permissions? Should only the algorithm editors be able to view this?

Reply:
No, this is also used if the user's request has been rejected. In that case they can request permission again, changing the status back to `pending`. If the user does not have editing rights, the value for `status` is always set to `pending` if the requesting user is not already a user or an editor for the algorithm.